src/analysis/convex/minmax.lean

```diff
@@ -0,0 +1,139 @@
+-/
+lemma is_min_on.of_is_local_min_on_of_convex_on_Icc {f : ℝ → ℝ} {a b : ℝ}
+  (a_lt_b : a < b) (h_local_min : is_local_min_on f (Icc a b) a) (h_conv : convex_on (Icc a b) f) :
+  ∀ x ∈ Icc a b, f a ≤ f x :=
+begin
+  by_contradiction H_cont,
+  push_neg at H_cont,
+  rcases H_cont with ⟨x, ⟨x_in_I, fx_lt_fa⟩⟩,
+
+  have exists_nhd_min : ∃ z > a, ∀ y ∈ (Icc a z), f y ≥ f a,
+  { dsimp [is_local_min_on, is_min_filter] at h_local_min,
+    rw [eventually_iff_exists_mem] at h_local_min,
+    rcases h_local_min with ⟨U, U_in_nhds_within, fy_ge_fa⟩,
+    rw [nhds_within_Icc_eq_nhds_within_Ici a_lt_b,mem_nhds_within_Ici_iff_exists_Icc_subset]
+        at U_in_nhds_within,
+    rcases U_in_nhds_within with ⟨ε, ε_in_Ioi, Ioc_in_U⟩,
+    refine ⟨ε, mem_Ioi.mp ε_in_Ioi, _⟩,
+    exact λ y y_in_Ioc, fy_ge_fa y (mem_of_mem_of_subset y_in_Ioc Ioc_in_U) },
+
+  rw [mem_Icc] at x_in_I,
+  rcases x_in_I with ⟨h_ax, h_xb⟩,
+
+  have a_ne_x : a ≠ x,
+  { by_contradiction H,
+    push_neg at H,
+    rw H at fx_lt_fa,
+    exact ne_of_lt fx_lt_fa rfl },
+
+  have a_lt_x : a < x := lt_of_le_of_ne h_ax a_ne_x,
+
+  have lt_on_nhd : ∀ y ∈ Ioc a x, f y < f a,
+  { intros y y_in_Ioc,
+    rcases (convex.mem_Ioc a_lt_x).mp y_in_Ioc with ⟨ya, yx, ya_pos, yx_pos, yax, y_combo⟩,
+    have h₁ : f (ya * a + yx * x) ≤ ya * f a + yx * f x :=
+      h_conv.2 (left_mem_Icc.mpr (le_of_lt a_lt_b)) ⟨h_ax, h_xb⟩ (ya_pos)
+        (le_of_lt yx_pos) yax,
+
+    rw [y_combo] at h₁,
+    calc
+      f y ≤ ya * f a + yx * f x       : h₁
+      ... < ya * f a + yx * f a       : by linarith [(mul_lt_mul_left yx_pos).mpr fx_lt_fa]
+      ... = f a                       : by rw [←add_mul, yax, one_mul] },
+
+  rcases exists_nhd_min with ⟨z, hz, ge_on_nhd⟩,
+  by_cases h_xz : x ≤ z,
+  { linarith [ge_on_nhd x (show x ∈ Icc a z, by exact ⟨h_ax, h_xz⟩)] },
+  { have h₁ : z ∈ Ioc a x := ⟨hz, le_of_not_ge h_xz⟩,
+    have h₂ : z ∈ Icc a z := ⟨le_of_lt hz, le_refl z⟩,
+    linarith [lt_on_nhd z h₁, ge_on_nhd z h₂] }
+end
+
+/--
+A local minimum of a convex function is a global minimum, restricted to a set `s`.
+-/
+lemma is_min_on.of_is_local_min_on_of_convex_on {f : E → ℝ} {a : E}
+  (a_in_s : a ∈ s) (h_localmin: is_local_min_on f s a) (h_conv : convex_on s f) :
+  ∀ x ∈ s, f a ≤ f x :=
+begin
+  by_contradiction H_cont,
+  push_neg at H_cont,
+  rcases H_cont with ⟨x, ⟨x_in_s, fx_lt_fa⟩⟩,
+
+  let g : affine_map ℝ ℝ ℝ E E := affine_map.line_map a (x - a),
+
+  have hg0 : g 0 = a := affine_map.line_map_apply_zero a (x - a),
+  have hg1 : g 1 = x := by simp [affine_map.line_map_apply, one_smul],
+
+  have fg_local_min_on : is_local_min_on (f ∘ g) (g ⁻¹' s) 0,
+  { rw [←hg0] at h_localmin,
+    refine is_local_min_on.comp_continuous_on h_localmin subset.rfl
+      (continuous.continuous_on (affine_map.line_map_continuous)) _,
+    simp [mem_preimage, hg0, a_in_s] },
+
+  have fg_min_on : is_min_on (f ∘ g) (Icc 0 1) 0,
```

Comment:
Since you will never use this fact without unfolding the definition of `is_min_on` first, why not directly stating `have fg_min_on : ∀ x ∈ (Icc 0 1 : set ℝ), (f ∘ g) 0 ≤ (f ∘ g) x,`?

Reply:
Yes -- changed.